### PR TITLE
Fix Half comparison

### LINKED ISSUE #39770: `System.Half` doesn't generate correct ordered comparison results when both sides are negative
### Description
The comparison methods / operators do not return the correct comparison result when both numbers are negative. For example, `-1 <= NegativeInfinity` returns `true`, when it is obvious that -1 is greater than `NegativeInfinity`.

### Configuration

The code likely isn't specific to a certain H/W or OS. There's only software implementations of those comparison methods.

### Regression?

This bug existed from the initial prototype code in dotnet/corefxlab.

### Other information

The current implementation, `return (short)(left._value) <= (short)(right._value); ` is incorrect when both sides are negative.

Consider the following:
|Actual value|Raw bits            | Signed raw bit comparison | Unsigned raw bit comparison|Half comparison|
|------------|--------------------|---------------------------|-------------------------|---------------|
|          -1|`1 01111 0000000000`| Smaller (-17408)          | Smaller (48128)         | Larger        |
|          -∞|`1 11111 0000000000`| Larger (-1024)            | Larger  (64512)         | Smaller       |
|------------|--------------------|---------------------------|-------------------------|---------------|
|           1|`0 01111 0000000000`| Smaller (15360)           | Smaller (15360)         | Smaller       |
|           ∞|`0 11111 0000000000`| Larger  (31744)           | Larger  (31744)         | Larger        |

This implementation is correct for cases where both sides are positive, but when both sides are negative we get an inverted result. This likely is the same in all 4 ordered comparison operators and `CompareTo` method.

The solution would be to invert the result when `leftIsNegative` is true. (XOR with `leftIsNegative`?)

https://github.com/dotnet/runtime/blob/0e23ef47ff8a2de6a5b5898a849ba00303aba8df/src/libraries/System.Private.CoreLib/src/System/Half.cs#L127-L145

P.S. Apparently we don't have any test that compares two negative values....

### PR DESCRIPTION
**Pull request by Gnbrkm41**
Fixes #39770

* Fixes the comparison to be correct when both numbers are negative
* Add relevant tests

cc @tannergooding

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Half.cs [hunk 1]
```diff
@@ -141,7 +141,7 @@ private ushort Significand
                 // says they should be equal, even if the signs differ.
                 return leftIsNegative || AreZero(left, right);
             }
-            return (short)(left._value) <= (short)(right._value);
+            return left._value <= right._value ^ leftIsNegative;
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Half.cs:144]: Explicit parentheses would be nice, to ensure this is doing the right thing: https://github.com/ucb-bar/berkeley-softfloat-3/blob/master/source/f16_le_quiet.c

File: src/libraries/System.Private.CoreLib/src/System/Half.cs [hunk 2]
```diff
@@ -116,7 +116,7 @@ private ushort Significand
                 // says they should be equal, even if the signs differ.
                 return leftIsNegative && !AreZero(left, right);
             }
-            return (short)(left._value) < (short)(right._value);
+            return left._value < right._value ^ leftIsNegative;
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Half.cs:119]: Just for reference, here is the lt implementation: https://github.com/ucb-bar/berkeley-softfloat-3/blob/master/source/f16_lt_quiet.c

File: src/libraries/System.Runtime/tests/System/HalfTests.cs [hunk 3]
```diff
@@ -281,6 +281,10 @@ public static IEnumerable<object[]> CompareTo_TestData()
             yield return new object[] { Half.NaN, Half.NaN, 0 };
             yield return new object[] { Half.NaN, UInt16BitsToHalf(0x0000), -1 };
             yield return new object[] { Half.MaxValue, null, 1 };
+            yield return new object[] { Half.MinValue, Half.NegativeInfinity, 1 };
```
Comment (tannergooding) [inline src/libraries/System.Runtime/tests/System/HalfTests.cs:284]: Are the same tests missing from float/double? It might be nice to add them if so.

Comment (Gnbrkm41) [reply src/libraries/System.Runtime/tests/System/HalfTests.cs:284]: Apparently we do. 👀

File: src/libraries/System.Private.CoreLib/src/System/Half.cs (hunk 2 above)
Comment (Gnbrkm41) [reply src/libraries/System.Private.CoreLib/src/System/Half.cs:119]: Looks like it's pretty much the same.

**Comment by tannergooding** [comment]
@tarekgh, @GrabYourPitchforks; seems to be a number of ICU failures that are likely unrelated.

**Comment by tarekgh** [comment]
@safern CI started to get the following failures, any idea what has changed?

```C#
Executed on dci-mac-build-043.local
+ ./RunTests.sh --runtime-path /tmp/helix/working/B6540A0D/p
----- start Thu Jul 23 02:11:51 PDT 2020 =============== To repro directly: =====================================================
pushd .
/tmp/helix/working/B6540A0D/p/dotnet exec --runtimeconfig Common.Tests.runtimeconfig.json --depsfile Common.Tests.deps.json xunit.console.dll Common.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing 
popd
===========================================================================================================
/private/tmp/helix/working/B6540A0D/w/AC97099A/e /private/tmp/helix/working/B6540A0D/w/AC97099A/e
Process terminated. Couldn't find a valid ICU package installed on the system. Set the configuration flag System.Globalization.Invariant to true if you want to run with no globalization support.
   at System.Environment.FailFast(System.String)
   at System.Globalization.GlobalizationMode.GetGlobalizationInvariantMode()
   at System.Globalization.GlobalizationMode..cctor()
   at System.Globalization.GlobalizationMode.get_Invariant()
   at System.Globalization.CultureData.CreateCultureWithInvariantData()
   at System.Globalization.CultureData.get_Invariant()
   at System.Globalization.CultureInfo..cctor()
   at System.Globalization.CultureInfo.get_CachedCulturesByName()
   at System.Globalization.CultureInfo.GetCultureInfo(System.String)
   at System.Reflection.RuntimeAssembly.GetLocale()
   at System.Reflection.RuntimeAssembly.GetName(Boolean)
   at System.Reflection.Assembly.GetName()
   at Internal.Microsoft.Extensions.DependencyModel.DependencyContextLoader.LoadAssemblyContext(System.Reflection.Assembly, Internal.Microsoft.Extensions.DependencyModel.IDependencyContextReader)
   at Internal.Microsoft.Extensions.DependencyModel.DependencyContextLoader.Load(System.Reflection.Assembly)
   at Internal.Microsoft.Extensions.DependencyModel.DependencyContext.Load(System.Reflection.Assembly)
   at Xunit.AssemblyHelper..ctor(System.String, Xunit.Abstractions.IMessageSink)
   at Xunit.AssemblyHelper.SubscribeResolveForAssembly(System.Type, Xunit.Abstractions.IMessageSink)
   at Xunit.ConsoleClient.Program.Main(System.String[])
```

**Comment by tarekgh** [comment]
CC @akoeplinger for the ICU failures on OSX if this is related to #39833

**Comment by safern** [comment]
> CC @akoeplinger for the ICU failures on OSX if this is related to #39833

@tarekgh I just validated and the build didn't not have @akoeplinger changes. I will trigger a build again.

**Review by tannergooding — APPROVED** [review]

**Pull request merged**